src/audio/manager.cpp

```diff
@@ -130,70 +130,125 @@ AudioManager::AudioManager(const Glib::ustring &backends_string)
+    RNNoiseUninitialize();
+#endif
+}
 
-        if (auto code = ma_device_init(&m_context, &m_capture_config, &m_capture_device); code != MA_SUCCESS) {
-            spdlog::get("audio")->error("failed to initialize capture device (code: {})", static_cast<int>(code));
-            m_ok = false;
-            return;
-        }
+void AudioManager::OpenPlaybackDevice(const ma_device_id device_id) {
+    assert(!m_playback_device_ready && "Tried to open new playback device without closing the current one");
+
+    auto config = ma_device_config_init(ma_device_type_playback);
+    config.playback.format = ma_format_f32;
+    config.playback.channels = 2;
+    config.playback.pDeviceID = &device_id;
+    config.sampleRate = 48000;
+    config.dataCallback = data_callback;
+    config.pUserData = this;
 
-        char capture_device_name[MA_MAX_DEVICE_NAME_LENGTH + 1];
-        ma_device_get_name(&m_capture_device, ma_device_type_capture, capture_device_name, sizeof(capture_device_name), nullptr);
-        spdlog::get("audio")->info("using {} as capture device", capture_device_name);
+    auto result = ma_device_init(&m_context, &config, &m_playback_device);
+    if (result != MA_SUCCESS) {
+        spdlog::get("audio")->error("Failed to initialize playback device (code: {})", static_cast<int>(result));
+        return;
     }
 
-    Glib::signal_timeout().connect(sigc::mem_fun(*this, &AudioManager::DecayVolumeMeters), 40);
+    result = ma_device_start(&m_playback_device);
+    if (result != MA_SUCCESS) {
+        spdlog::get("audio")->error("Failed to start playback device (code: {})", static_cast<int>(result));
+
+        ma_device_uninit(&m_playback_device);
+        return;
+    }
+
+    m_playback_device_ready = true;
 }
 
-AudioManager::~AudioManager() {
+void AudioManager::OpenCaptureDevice(const ma_device_id device_id) {
+    assert(!m_capture_device_ready && "Tried to open new capture device without closing the current one");
+
+    auto config = ma_device_config_init(ma_device_type_capture);
+    config.capture.format = ma_format_s16;
+    config.capture.channels = 2;
+    config.capture.pDeviceID = &device_id;
+    config.sampleRate = 48000;
+    config.periodSizeInFrames = 480;
+    config.dataCallback = capture_data_callback;
+    config.pUserData = this;
+
+    auto result = ma_device_init(&m_context, &config, &m_capture_device);
+    if (result != MA_SUCCESS) {
+        spdlog::get("audio")->error("Failed to initialize capture device (code: {})", static_cast<int>(result));
+        return;
+    }
+
+    result = ma_device_start(&m_capture_device);
+    if (result != MA_SUCCESS) {
+        spdlog::get("audio")->error("Failed to start capture device (code: {})", static_cast<int>(result));
+
+        ma_device_uninit(&m_capture_device);
+        return;
+    }
+
+    m_capture_device_ready = true;
+}
+
+void AudioManager::TryOpenPlaybackDevice(const ma_device_id device_id) {
+    OpenPlaybackDevice(std::move(device_id));
```

Comment:
std::move doesnt do anything on these structs so they can just be gotten rid of

Reply:
Not very experienced in C++, I used move here to ensure that the struct doesn't get copied around.

Reply:
it gets copied anyways so it doesnt do anything. also double doesnt do anything because you declare it const (not that it would without the const either). i would just remove them

Reply:
I suppose I could just pass a reference then to avoid the copy?